src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubAppCredentials.java

```diff
@@ -0,0 +1,184 @@
+import com.cloudbees.plugins.credentials.common.StandardUsernamePasswordCredentials;
+import com.cloudbees.plugins.credentials.impl.BaseStandardCredentials;
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.Extension;
+import hudson.Util;
+import hudson.util.FormValidation;
+import hudson.util.ListBoxModel;
+import hudson.util.Secret;
+import java.io.IOException;
+import java.util.List;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.github.GHApp;
+import org.kohsuke.github.GHAppInstallation;
+import org.kohsuke.github.GHAppInstallationToken;
+import org.kohsuke.github.GitHub;
+import org.kohsuke.github.GitHubBuilder;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.DataBoundSetter;
+import org.kohsuke.stapler.QueryParameter;
+import org.kohsuke.stapler.verb.POST;
+
+import static org.jenkinsci.plugins.github_branch_source.GitHubSCMNavigator.DescriptorImpl.getPossibleApiUriItems;
+import static org.jenkinsci.plugins.github_branch_source.JwtHelper.createJWT;
+
+public class GitHubAppCredentials extends BaseStandardCredentials implements StandardUsernamePasswordCredentials {
+
+    private static final String ERROR_AUTHENTICATING_GITHUB_APP = "Couldn't authenticate with GitHub app ID %s";
+
+    @NonNull
+    private final String appID;
+
+    @NonNull
+    private final Secret privateKey;
+
+    private String apiUri;
+
+    @DataBoundConstructor
+    @SuppressWarnings("unused") // by stapler
+    public GitHubAppCredentials(
+            CredentialsScope scope,
+            String id,
+            @CheckForNull String description,
+            @NonNull String appID,
+            @NonNull Secret privateKey
+    ) {
+        super(scope, id, description);
+        this.appID = appID;
+        this.privateKey = privateKey;
+    }
+
+    public String getApiUri() {
+        return apiUri;
+    }
+
+    @DataBoundSetter
+    public void setApiUri(String apiUri) {
+        this.apiUri = apiUri;
+    }
+
+    @NonNull
+    public String getAppID() {
+        return appID;
+    }
+
+    @NonNull
+    public Secret getPrivateKey() {
+        return privateKey;
+    }
+
+    @SuppressWarnings("deprecation") // preview features are required for GitHub app integration, GitHub api adds deprecated to all preview methods
+    static String generateAppInstallationToken(String appId, String appPrivateKey, String apiUrl) {
+        try {
+            String jwtToken = createJWT(appId, appPrivateKey);
+            GitHub gitHubApp = new GitHubBuilder().withEndpoint(apiUrl).withJwtToken(jwtToken).build();
+
+            GHApp app = gitHubApp.getApp();
+
+            List<GHAppInstallation> appInstallations = app.listInstallations().asList();
+            if (!appInstallations.isEmpty()) {
+                GHAppInstallation appInstallation = appInstallations.get(0);
```

Comment:
I guess you are assuming the app is installed into exactly one org? Presumably this code would not work if it were installed into multiple orgs and you were trying to use the same credentials to authenticate to multiple org folders.

Reply:
yes, should be pretty easy to add multiple org support on top of this.

Reply:
If this is a limitation (1 GitHub App for 1 GitHub organization), I suggest we document it clearly. My usage pattern was to install the same GitHub App across many GitHub orgs. Unless the multi org support is coming very soon.

Reply:
Particularly given https://github.com/jenkinsci/github-branch-source-plugin/pull/269#discussion_r395872974 it seems to me that the app is pretty much only a formality to get suitable credentials, so there is no particular advantage to using one app across multiple orgs other than quicker initial setup. Either way, clearer documentation would be appreciated.

Reply:
> there is no particular advantage to using one app across multiple orgs other than quicker initial setup

It's a good point. I am not sure what is the best model. I was thinking that it would be best to have one app for one Jenkins, and many installations. This is how other CIs ([Travis](https://github.com/travis-ci/travis_migrate_to_apps), [Google Cloud builds](https://cloud.google.com/cloud-build/docs/automating-builds/run-builds-on-github)) do it.

Creating a GitHub app for each instantiation means:

* loosing the ability to control permissions in one single app. If permission needs to be updated (for example to provide checks api access) then multiple apps will have to be updated where GitHub App permission updates are nicely managed with an approval flow
* an app can be installed at the level of a GitHub org, or GitHub repository. That means many GitHub apps to create and maintain

On the other end, one app per installation may also mean better security: you can authorize write access to the GitHub content in one app and prohibit it in another.

Supporting multiple installations would support all options. If it can't be done or hard to do, it would be best to warn that multiple installations have been detected, and that is currently not supported.

Note that I don't think this should prevent the release.

Reply:
> This is how other CIs … do it.

But those are SaaS’s, so the creator of the singleton app would be the development/operations team for the CI; whereas in the case of Jenkns, the individual administrator must create their own (perhaps “internal”) app.

> Supporting multiple installations would support all options.

Yes, it would be desirable to support this mode. Looks like it would be tricky, since the internal (Java) API calls to get the token do not supply any context that would let the plugin determine which organization is making the request. I guess we could use some `ThreadLocal` hack.

Reply:
It was pretty straightforward, all context needs to be in the credential, it does mean some duplication currently (api url has the same issue), but in the future the job configuration could be simplified probably.

pull request here:
https://github.com/jenkinsci/github-branch-source-plugin/pull/290